### ISSUE TITLE
SNOW-1017272: Cannot install snowflake-connector-python with cryptography>=42.0.0

### ISSUE THREAD

**Issue by sodul** [snowflakedb/snowflake-connector-python#1862]
### Python version

Python 3.12.1 (main, Dec 18 2023, 11:07:36) [Clang 15.0.0 (clang-1500.0.40.1)]

### Operating system and processor architecture

macOS-14.3-arm64-arm-64bit

### Installed packages

```python
N/A (the list is not relevant)
```

### What did you do?

```python
> pip install snowflake-connector-python==3.6.0 cryptography==42.0.0
Requirement already satisfied: snowflake-connector-python==3.6.0 in /Users/stephane/.pyenv/versions/3.12.1/lib/python3.12/site-packages (3.6.0)
Collecting cryptography==42.0.0
  Using cached cryptography-42.0.0-cp39-abi3-macosx_10_12_universal2.whl.metadata (5.3 kB)
Requirement already satisfied: asn1crypto<2.0.0,>0.24.0 in /Users/stephane/.pyenv/versions/3.12.1/lib/python3.12/site-packages (from snowflake-connector-python==3.6.0) (1.5.1)
Requirement already satisfied: cffi<2.0.0,>=1.9 in /Users/stephane/.pyenv/versions/3.12.1/lib/python3.12/site-packages (from snowflake-connector-python==3.6.0) (1.16.0)
INFO: pip is looking at multiple versions of snowflake-connector-python to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install cryptography==42.0.0 and snowflake-connector-python==3.6.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested cryptography==42.0.0
    snowflake-connector-python 3.6.0 depends on cryptography<42.0.0 and >=3.1.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

### What did you expect to see?

Both packages should install.

**Comment by sodul** [comment]
Note that we have the same issue with `pyOpenSSL==24.0.0`.

**Comment by sfc-gh-aling** [comment]
thanks for the report! we will bump the version and release a newer version after verifying the dependencies.
PR is here: https://github.com/snowflakedb/snowflake-connector-python/pull/1866

**Comment by sodul** [comment]
@sfc-gh-aling May I suggest to drop the upper versions limits for dependencies? Each time there is a new release of pyOpenSSL that other third party packages use such as cryptography or moto we risk having to hold back until the snowflake-connector is updated to remove this artificial limitation.

If there is a know compatibility issue, that would make sense but the preemptive limits are causing more grief than necessary.

**Comment by bbayles** [comment]
[Here](https://github.com/snowflakedb/snowflake-connector-python/issues/1629#issuecomment-1636169207) is a previous note on the same topic. I think the logic still applies.

**Comment by sfc-gh-aling** [comment]
thanks for all your feedbacks. want to share some updates with you:

- we will first do a release of connector bumping the cryptography dependency next week.
- I am driving the discussion with the team on relaxing the cryptography dependencies. our team is onboard with the idea to unpin cryptography upper bound, and we've identified that a more robust infra is what we need to guarantee that cryptography doesn't break us or we're able to react to any dependency issue asap. we have identified action items and will work on it.

**Comment by sodul** [comment]
One strategy we have applied is to trick `pip` to ignore the dependency requirements by having a custom script that will call `pip` multiple times with a generated constraint file. One of the worst offenders in our case is the azure cli so we install it separately, and feed our requirements file as a constraint file (replacing `==` entries with `>=`), then calling pip again with the real requirements file. By having the picky packages installed first it tricks pip to install conflicting requirements.

This is dirty and makes our `pip` installations less efficient but without this we would not be able to install our large set of packages. We keep this reasonably safe by having good test coverage on our python code base to ensure nothing actually breaks at runtime.

**Comment by anentropic** [comment]
I can see this has already been fixed in a PR here, but we are all waiting on the release now.

Are you sure that strictly upper-bounding the cryptography lib version is the best strategy?  It makes your downstream lib the bottleneck for your users getting Dependabot alerts in their projects.

`cryptography` lib claims to have a "strong API stability" policy even across major semver versions https://cryptography.io/en/latest/api-stability/ It seems like exactly the kind of lib where you'd always want to be on the latest version. 

Or could you consider having some kind of fast path on your side for pushing out security updates? This issue has been languishing for several weeks now.

**Comment by DellWard** [comment]
Hi, any update on the mentioned upcoming release for this week? When should we expect it?

**Comment by baruchoxman** [comment]
Note that cryptography 42.0.4 was just released. It addresses CVE-2024-26130. Please try including this version in the cryptography version dependency bump.

**Comment by sfc-gh-aling** [comment]
thanks for your patience, we're working on the release at the moment. it should be available late today or tomorrow.
the newer version will have `cryptography>=3.1.0,<43.0.0`.

**Comment by anentropic** [comment]
will you have a procedure in place for faster turnaround when this happens again with v43 ?

**Comment by bbayles** [comment]
It's good that a new release will be available, but merely going up to `43.0.0` is really missing the point!

Many software projects use a `Major.Minor.Patch` versioning scheme (semver). The `Major` version incrementing indicates that something is expected to break, and that work will need to be done to port. __cryptography__ is not such a project, and applying that logic to it is incorrect.

The document [here](https://cryptography.io/en/latest/api-stability/#what-does-stable-mean) describes the versioning and API stability policy for `cryptography`. It notes that the [old policy](https://cryptography.io/en/latest/api-stability/#previous-scheme) matched semver, but the new one doesn't.

Per a previous comment:

> The risk is that there is a breaking change in a `cryptography` upgrade (i.e. one that's against their stated policy) that affects applications that blindly upgrade. However, I think this is a risk worth taking:
> * Applications shouldn't be blindly upgrading either `cryptography` or `snowflake-connector-python`!
> * If there's some fix so important that it requires a breaking change in `cryptography`, it's likely a very serious security issue that ought to break loudly

I'm afraid that right now `snowflake-connector-python`'s policy is aiming for compatibility at the expense of security, and is failing to deliver even on compatibility.

**Comment by sodul** [comment]
The `cryptography>=42.0.0` issue is fixed in v3.7.1, but the underlying issue of pre-emptively locking out unreleased versions of dependencies remains.
